File: netflix-frontend/src/services/watchHistoryService.js
import api from './api';

const watchHistoryService = {
  getWatchHistory: async (page = 0, size = 20) => {
    const response = await api.get('/history', { params: { page, size } });
    return response.data;
  },

  getContinueWatching: async () => {
    const response = await api.get('/history/continue-watching');
    return response.data;
  },

  saveWatchProgress: async (movieId, progressSeconds, totalDurationSeconds, quality) => {
    const response = await api.post(`/history/${movieId}/progress`, {
      progressSeconds,
      totalDurationSeconds,
      quality
    });
    return response.data;
  },

  getWatchProgress: async (movieId) => {
    const response = await api.get(`/history/${movieId}`);
    return response.data;
  },

  deleteWatchHistory: async (movieId) => {
    const response = await api.delete(`/history/${movieId}`);
    return response.data;
  },

  validateStreamingAccess: async (movieId, quality = null) => {
    const params = quality ? { quality } : {};
    const response = await api.get(`/streaming/${movieId}/validate`, { params });
    return response.data;
  },

  getStreamUrl: async (movieId, quality = 'SD') => {
    const response = await api.get(`/streaming/${movieId}/stream`, { params: { quality } });
    return response.data;
  }
};

export default watchHistoryService;

File: netflix-backend/src/main/java/com/netflix/streaming/service/WatchHistoryService.java
package com.netflix.streaming.service;

import com.netflix.streaming.dto.request.WatchProgressRequest;
import com.netflix.streaming.model.WatchHistory;
import com.netflix.streaming.repository.WatchHistoryRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class WatchHistoryService {

    private final WatchHistoryRepository watchHistoryRepository;

    public WatchHistory saveWatchProgress(String userId, String movieId, WatchProgressRequest request) {
        WatchHistory history = watchHistoryRepository.findByUserIdAndMovieId(userId, movieId)
                .orElse(new WatchHistory());

        if (history.getId() == null) {
            history.setUserId(userId);
            history.setMovieId(movieId);
            history.setFirstWatchedAt(LocalDateTime.now());
            history.setWatchCount(1);
        } else {
            history.setWatchCount(history.getWatchCount() + 1);
        }

        history.setProgressSeconds(request.getProgressSeconds());
        history.setTotalDurationSeconds(request.getTotalDurationSeconds());
        history.setWatchedQuality(request.getQuality());
        history.setLastWatchedAt(LocalDateTime.now());

        double watchPercentage = (double) request.getProgressSeconds() / request.getTotalDurationSeconds() * 100;
        history.setCompleted(watchPercentage >= 90);

        return watchHistoryRepository.save(history);
    }

    public Page<WatchHistory> getWatchHistory(String userId, int page, int size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by("lastWatchedAt").descending());
        return watchHistoryRepository.findByUserId(userId, pageable);
    }

    public List<WatchHistory> getContinueWatching(String userId) {
        return watchHistoryRepository.findByUserIdAndCompletedFalseOrderByLastWatchedAtDesc(userId);
    }

    public void deleteWatchHistory(String userId, String movieId) {
        watchHistoryRepository.deleteByUserIdAndMovieId(userId, movieId);
    }

    public WatchHistory getWatchProgress(String userId, String movieId) {
        return watchHistoryRepository.findByUserIdAndMovieId(userId, movieId)
                .orElse(null);
    }
}
